libtransmission/variant.cc

```diff
@@ -737,80 +739,79 @@ bool tr_variantDictRemove(tr_variant* dict, tr_quark const key)
 ****  BENC WALKING
 ***/
 
-struct KeyIndex
+class WalkNode
 {
-    char const* keystr;
-    tr_variant* val;
-};
-
-static int compareKeyIndex(void const* va, void const* vb)
-{
-    auto const* const a = static_cast<struct KeyIndex const*>(va);
-    auto const* const b = static_cast<struct KeyIndex const*>(vb);
-
-    return strcmp(a->keystr, b->keystr);
-}
+private:
+    // When walking `v`'s children, this is the index of the next child
+    size_t child_index = 0;
 
-struct SaveNode
-{
-    tr_variant const* v;
-    tr_variant* sorted;
-    size_t childIndex;
-    bool isVisited;
-};
-
-static void nodeConstruct(struct SaveNode* node, tr_variant const* v, bool sort_dicts)
-{
-    node->isVisited = false;
-    node->childIndex = 0;
+    // When `v` is a dict, this is its children's indices sorted by key.
+    // Bencoded dicts must be sorted, so this is useful when writing benc.
+    std::vector<size_t> sorted;
 
-    if (sort_dicts && tr_variantIsDict(v))
+    void sortByKey()
     {
-        /* make node->sorted a sorted version of this dictionary */
+        auto const n = v.val.l.count;
 
-        size_t const n = v->val.l.count;
-        struct KeyIndex* tmp = tr_new(struct KeyIndex, n);
+        struct ByKey
+        {
+            char const* key;
+            size_t idx;
+        };
 
-        for (size_t i = 0; i < n; i++)
+        auto const* children = v.val.l.vals;
+        auto tmp = std::vector<ByKey>(n);
+        for (size_t i = 0; i < n; ++i)
         {
-            tmp[i].val = v->val.l.vals + i;
-            tmp[i].keystr = tr_quark_get_string(tmp[i].val->key, nullptr);
+            tmp[i] = { tr_quark_get_string(children[i].key, nullptr), i };
         }
 
-        qsort(tmp, n, sizeof(struct KeyIndex), compareKeyIndex);
+        auto const compare = [](ByKey const& a, ByKey const& b)
+        {
+            return strcmp(a.key, b.key) < 0;
+        };
+        std::sort(std::begin(tmp), std::end(tmp), compare);
 
-        node->sorted = tr_new(tr_variant, 1);
-        tr_variantInitDict(node->sorted, n);
+        //  keep the sorted indices
 
+        sorted.resize(n);
         for (size_t i = 0; i < n; ++i)
         {
-            node->sorted->val.l.vals[i] = *tmp[i].val;
+            sorted[i] = tmp[i].idx;
         }
+    }
 
-        node->sorted->val.l.count = n;
+public:
+    bool is_visited = false;
 
-        tr_free(tmp);
+    // Shallow bitwise copy of the variant passed to the constructor
+    tr_variant const v = {};
 
-        v = node->sorted;
-    }
-    else
+    WalkNode(tr_variant const* v_in, bool sort_dicts)
+        : v{ *v_in }
     {
-        node->sorted = nullptr;
+        if (sort_dicts && tr_variantIsDict(v_in))
+        {
+            sortByKey();
+        }
     }
 
-    node->v = v;
-}
+    tr_variant const* nextChild()
+    {
+        if (child_index >= v.val.l.count)
```

Comment:
Feeling a bit uneasy using `l` when it might not be valid. True that we're not calling this method (at the moment) unless `v` is a container, but as a standalone class I'd welcome it to have an assert here just in case, or even a proper check and a `nullptr` result, when `v` isn't a container after all.

Reply:
good catch. Safeguard added at ba418d1